src/aws-lambda/personalize-pre-create-campaigns/personalize-pre-create-campaigns.py

```diff
@@ -284,526 +657,499 @@ def create_recent_purchase_filter(dataset_group_arn, ssm_parameter_name):
-                return {
-                    'statusCode': 200,
-                    'body': json.dumps('Rerank campaign status is NOT active')
-                }
-                
-            rerank_arn = campaign['campaignArn']
-
-            # Finally, set the campaign arn as the system parameter expected by services
-            logger.info("Setting rerank campaignArn " + rerank_arn + " as system parameter")
-            response = ssm.put_parameter(
-                Name=rerank_campaign_arn_param,
-                Description='Retail Demo Store Personalized Reranking Campaign Arn Parameter',
-                Value='{}'.format(rerank_arn),
-                Type='String',
-                Overwrite=True
-            )
-    else:
-        logger.info('Creating personalized reranking campaign')
-        personalize.create_campaign(
-            name = rerank_campaign_name,
-            solutionVersionArn = rerank_solution_version_arn,
-            minProvisionedTPS = 1
-        )
-        return {
-            'statusCode': 200,
-            'body': json.dumps('Initiated create personalized reranking campaign; will check back for status')
-        }
 
-    # If we get here... all campaigns have been successfully created!
-    logger.info('Related product, Product recommendation, and Personalized reranking campaigns fully provisioned!')
+def lambda_handler(event, context):
+    """According to the contents of the SSM variable retaildemostore-training-config build and delete
+    dataset groups, datasets, solutions and campaigns to reach the desired state."""
+    logger.debug('## ENVIRONMENT VARIABLES')
+    logger.debug(os.environ)
+    logger.debug('## EVENT')
+    logger.debug(event)
 
-    # No need for this lambda function to be called anymore so delete CW event rule that has been calling us.
-    delete_event_rule(lambda_event_rule_name)
+    ## Inspect the event - if it is coming from EventBridge then it is polling after reset
+    ## If it is coming from CloudFormation then use the helper to create or tear down
+    if "source" in event and event["source"] == "aws.events":
+        done = update()  # Will pick up desired behaviour from training_config
+        if done:
+            disable_event_rule(lambda_event_rule_name)
+            return {
+                'statusCode': 200,
+                'body': json.dumps("Event rule disabled")
+            }
+        else:
+            return {
+                'statusCode': 200,
+                'body': json.dumps("Polled Personalize Create Update")
+            }
 
-    return {
-        'statusCode': 200,
-        'body': json.dumps('Related product, Product recommendation, and Personalized reranking campaigns fully provisioned!')
-    }
\ No newline at end of file
+    else:
+        # We have a Cloud Formation Event - the event contains info about whether it wants to update, create or delete
+        cloudformation_helper(event, context)
```

Comment:
Changing this function to a custom resource is going to change the CFN deployment time from 25-30 minutes to 2+ hours (if enabled), right? I have concerns about the customer experience with deploying this project for developers as this length of time for a deployment would be very unusual. 

Also, this is going to break deployments using our (AWS) internal tool for workshops that will timeout deployments that take this long.

Reply:
We agree that 2+ hours would be too long to wait for CloudFormation, though this is not quite what happens. 

As was the case previously, once Personalize starts building, CloudFormation sees it as deployed. Indeed, if we allowed CloudFormation to wait for the Custom Resource Lambda to finish polling, CloudFormation itself would time out. So what we do in the Custom Resource Lambda is return immediately and post the event rule as was the case with the previous solution. CloudFormation finishes its deployment and Personalize continues to train by polling the Lambda with the event rule (if Personalize deployment is enabled). 

That event rule is also used for "reset button" functionality to poll the Lambda when changes are requested to the Personalize deployment.